libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_UTIL_FLAT_EXPR_TREE_HPP
+#define MAMBA_UTIL_FLAT_EXPR_TREE_HPP
+
+#include <functional>
+#include <utility>
+#include <variant>
+#include <vector>
+
+namespace mamba::util
+{
+    /**
+     * In array binary tree.
+     *
+     * A binary tree, each node is either a leaf, or a node with exactly two children.
+     * This data structure is light and nothing prevents the user from representing
+     * any kind of binary directed acyclic graph (e.g. there can be multiple trees,
+     * or nodes could have multiple parents).
+     */
+    template <typename Branch, typename Leaf>
+    class flat_binary_tree
+    {
+    public:
+
+        using branch_type = Branch;
+        using leaf_type = Leaf;
+        struct BranchNode
+        {
+            branch_type data;
+            std::size_t left_child = 0;
+            std::size_t right_child = 0;
+        };
+        using branch_node = BranchNode;
+        using leaf_node = leaf_type;
+        using node_type = std::variant<branch_node, leaf_node>;
+        using node_list = std::vector<node_type>;
+        using size_type = typename node_list::size_type;
+        using idx_type = size_type;
+
+        [[nodiscard]] auto size() const -> size_type;
+        [[nodiscard]] auto empty() const -> bool;
+
+        void clear();
+        void reserve(size_type size);
+
+        auto add_leaf(const leaf_type& leaf) -> idx_type;
+        auto add_leaf(leaf_type&& leaf) -> idx_type;
+        auto add_branch(const branch_type& branch, idx_type left_child, idx_type right_child)
```

Comment:
Side note C++23: `auto add_branch(this auto&& self, ...)` would have been useful here haha